Improve unit testing

Moddus opened the pull request:
Hello,

I would like to contribute to Qkan but not having unit testing support makes it hard as I don't want to break stuff :sos: . My idea was to use mocking for separate GUI components from business logic. The main idea is shown within the `TestM145Porter` class using the `mock_imports` function. Apparently, the project is using initialization with `__init__.py` files heavily which make mocking harder e.g. currently I'm 'stuck with:
```
File "C:\Users\Moddus\PycharmProjects\QKan\qkan\tools\dialogs\dbAdapt.py", line 23, in <module>
    class DbAdaptDialog(QKanDBDialog, QKanProjectDialog, FORM_CLASS_dbAdapt):  # type: ignore
TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases

```
where multi class inheritance is use by statically creating `FORM_CLASS_dbAdapt`.

As I'm not an expert, yet :smile: . Could you explain the need to importing using `__init__.py` files or would it be possible to use some additional layer to abstract? I'm not sure if the current structure is given by the QGIS plugin API. I inspect a few popular plugins but I couldn't find the same pattern as in Qkan.

Thanks for working on Qkan! Thanks for taking the time to read this :)

Happy to get suggestions/comments.

Moddus committed:
Improve unit testing

Hello,

I would like to contribute to Qkan but not having unit testing support makes it hard as I don't want to break stuff :sos: . My idea was to use mocking for separate GUI components from business logic. The main idea is shown within the `TestM145Porter` class using the `mock_imports` function. Apparently, the project is using initialization with `__init__.py` files heavily which make mocking harder e.g. currently I'm 'stuck with:
```  File "C:\Users\Moddus\PycharmProjects\QKan\qkan\tools\dialogs\dbAdapt.py", line 23, in <module>
    class DbAdaptDialog(QKanDBDialog, QKanProjectDialog, FORM_CLASS_dbAdapt):  # type: ignore
TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases```
where multi class inheritance is use by statically creating `FORM_CLASS_dbAdapt`.

As I'm not an expert, yet :smile: . Could you explain the need to importing using `__init__.py` files or would it be possible to use some additional layer to abstract? I'm not sure if the current structure is given by the QGIS plugin API. I inspect a few popular plugins but I couldn't find the same pattern as in Qkan.

Thanks for working on Qkan! Thanks for taking the time to read this :)

Happy to get suggestions/comments.

Invisi commented on test/mocking_helper.py:
```diff
@@ -0,0 +1,67 @@
+import sys
+from importlib import util
+from types import NoneType
```
We're still bound to Python 3.9+, so `NoneType` is [unfortunately unavailable](https://docs.python.org/3.10/library/types.html#types.NoneType). `None` should already be enough, though

Invisi commented on test/test_datacheck.py:
```diff
@@ -7,7 +7,7 @@
 sys.path.append(os.path.join(os.path.split(__file__)[0], ".."))
 
-from test import BASE_DATA, BASE_WORK, LOGGER, QgisTest, iface
+from base_test import BASE_DATA, BASE_WORK, LOGGER, QgisTest, iface
```
Big fan of moving stuff out of `__init__`, however this should read `.base_test` and be a relative import.

Invisi commented on test/wip/test_xml.py:
```diff
@@ -1,4 +1,4 @@
-from test import BASE_DATA, BASE_WORK, LOGGER, QgisTest, iface
+from test.base_test import BASE_DATA, BASE_WORK, LOGGER, QgisTest, iface
```
```suggestion
from ..base_test import BASE_DATA, BASE_WORK, LOGGER, QgisTest, iface
```

Invisi commented:
Sorry for the delay and thanks for your PR.
Unfortunately the underlying data for the tests contains real-life data and as such cannot be published. (Hence why `test/data` is empty)
As for the imports inside `__init__.py` of each sub-plugin, I don't see any code that's actually using it outside of the plugin loader in `qkan/__init__.py`. Could be changed to ease testing without any issues.

If you don't mind, I'd rebase this and merge after fixing the imports

Moddus commented:
Hey,

@Invisi also sorry for the delay! Thanks for having a look and your suggestions. I applied the following changes:
- Addressed your comments.
- Rebased to the qkan3 branch.
- Fixed meta class typing issues.

The current branch allows me to run the test `TestM145Porter`:
```
Launching unittests with arguments python -m unittest test_importer.TestM145Porter in C:\Users\Moddus\PycharmProjects\QKan\test\m145porter

mock: M145Porter


Ran 1 test in 0.002s

OK
```

I would be happy if you find time to review the current RP! FYI: The commit history is a bit messy now but I can `rebase`, if we have an agreement with the proposed approach.

Best,
Moddus

Moddus commented:
Hey,

I extended the unittest `TestM145Porter` a bit to show what is possible now.

Best,
Moddus

Invisi commented:
Hi there and again sorry for the delay, work had priority for a bit.
Unfortunately I've had to revert the metaclass commit as the addon won't load in qgis anymore with those metaclasses.

What I proposed, and not elaborated properly on (sorry), was the idea to just remove the imports for the applications from the submodules, resulting in the ability to just import anything without the dialogs being loaded.

As for `fzahl(text="20", n=2)` resulting in `20 / (10^2)`, that seems to be intended based on the usage in `_importSWMM.py`. `qkan.m145porter._import.fzahl` also doesn't seem to be used anywhere at all anyway :x
https://github.com/hoettges/QKan/blob/054ebacac216a66c390814ce2aeb0bb7402a3321/qkan/swmmporter/_importSWMM.py#L260-L261

In the current state this PR should be good to go from my side, unless you wish to change the test itself. Further tests for other modules should probably be separate PRs.

Happy New Year and thanks again :)

Moddus commented:
Happy New Year, and what a great way to kick it off!

>  remove the imports for the applications from the submodules, resulting in the ability to just import anything without the dialogs being loaded.

This makes perfect sense! Appreciate, your change adjusting the imports.

> Unfortunately I've had to revert the metaclass commit as the addon won't load in qgis anymore with those metaclasses.

Would you mind giving a quick summary how I can test a development version of QKan to test?

> Further tests for other modules should probably be separate PRs.

Exactly, was planning to add more once the PR is merged.

Again, thanks for the modifications to the PR would be happy if it could be merged.

Best,
Moddus

Invisi commented:
> Would you mind giving a quick summary how I can test a development version of QKan to test?

If by test you mean loading in QGIS, absolutely. On my machines I do that by symlinking the `qkan` folder inside the repo to `%appdata%\QGIS\QGIS3\profiles\default\python\plugins\qkan` and opening QGIS.
That does require uninstalling the regular version beforehand, of course.

```
# cmd
mklink /d %appdata%\QGIS\QGIS3\profiles\default\python\plugins\qkan qkan

# powershell
New-Item -Path $env:appdata\QGIS\QGIS3\profiles\default\python\plugins\qkan -ItemType SymbolicLink -Value qkan
```
Note that this requires an elevated shell, which you should be able open in the current folder by `shift`+right-clicking the background, and then selecting `Open in Terminal` while holding `ctrl`+`shift`. Otherwise just navigate to the repository manually.

---

I'll squash and merge the PR as soon as I've got the time to do so.

Moddus commented:
@Invisi thanks to merging! The next pull request will be more clean, I promise.
